click/_termui_impl.py

```diff
@@ -347,10 +350,12 @@ def _pipepager(text, cmd, color):
             break
 
 
-def _tempfilepager(text, cmd, color):
+def _tempfilepager(generator, cmd, color):
     """Page through text by invoking a program on a temporary file."""
     import tempfile
     filename = tempfile.mktemp()
+    # TODO: This never terminates if the passed generator never terminates.
+    text = "".join(generator)
```

Comment:
This could be grouped, but does not really help here, since the pager should get invoked only once:

```python
def grouper(iterable, n):
    while True:
        yield itertools.chain([next(iterable)],
                              itertools.islice(iterable, n-1))


for buf in grouper(generator, 500):
    ...
```

However, `islice` itself could be used to enforce some (very high) limit and/or some timeout mechanism could be put in place.

Reply:
what do you think about dropping OS/2 support, maybe dropping support for the PAGER environment variable on windows and using `more.exe`, and in case of emergency just falling back to writing to stdout?

Reason: I think the cleanest way would be to get rid of the temp file pager all together.

Reply:
Sounds good to me in general, but on the other hand it might be better to just fail here then instead in an easy way (given that there is support for it already).

Reply:
I think I didn't completely understand what you mean, under which circumstance do you think it's good to fail?

Reply:
Either like it's now (never ending), or by adding some timeout (e.g. 30 seconds).
The first option would be easier, but then it should be noted with this new feature.
Anyway, I am not really in the position to decide this.
/cc @mitsuhiko

Reply:
I think it would be better to stream the generator to a file instead of buffering the contents in memory (which is the result of `"".join()`). This way you'd support large generators without using a lot of memory.

I cannot think of a proper solution for infinite input and buffering to a temporary file.

Reply:
I think I'll try to find a solution without tempfile for windows maybe on sunday, and for OS2 I'll change the code to use `_nullpager` (which is already a fallback in some cases). Then I can remove `_tempfilepager` hopefully :)